File: mesh/router/sovereign-ast-matrix-py/router_matrix.py
from __future__ import annotations
import queue
import threading
import time
from router_config import DB, FIFO_MAX, PROVIDERS, STICKY_TTL
from router_health import HealthDB




# ---------------------------------------------------------------------------
# State: ELO, circuit breakers, sticky, FIFO, health DB
# ---------------------------------------------------------------------------
class Matrix:
    def __init__(self) -> None:
        self.lock: threading.Lock = threading.Lock()
        self.fail: dict[str, tuple[int, float]] = {}
        self.elo: dict[str, float] = {p: 1000.0 for p in PROVIDERS}
        self.circuit: dict[str, str] = {p: "closed" for p in PROVIDERS}
        self.circuit_open_until: dict[str, float] = {}
        self.fifo: queue.Queue[int] = queue.Queue(maxsize=FIFO_MAX)
        self.health = HealthDB(DB)

    def record(
        self,
        model: str,
        prov: str,
        status: int,
        lat: float,
        winner: int = 0,
        strategy: str = "",
        session: str = "",
    ) -> None:
        lat_ms = lat * 1000
        with self.lock:
            self.health.record_request(
                prov,
                model,
                status,
                lat_ms,
                strategy,
                winner,
                session,
            )
            if status == 200:
                old_circuit = self.circuit.get(prov, "closed")
                self.elo[prov] = self.elo.get(prov, 1000) + 16
                self.fail[prov] = (0, time.time())
                self.circuit[prov] = "closed"
                if old_circuit != "closed":
                    self.health.record_healing(
                        prov,
                        model,
                        "circuit_recovered",
                        prev_status=old_circuit,
                        new_status="closed",
                    )
            elif status == 429:
                self.health.record_rate_limit(prov, model, status)
                self.elo[prov] = max(100.0, self.elo.get(prov, 1000) - 8)
                c, _ = self.fail.get(prov, (0, 0))
                self.fail[prov] = (c + 1, time.time())
            else:
                c, _ = self.fail.get(prov, (0, 0))
                self.fail[prov] = (c + 1, time.time())
                self.elo[prov] = max(100.0, self.elo.get(prov, 1000) - 32)
                if c + 1 >= 3:
                    old = self.circuit.get(prov, "closed")
                    self.circuit[prov] = "open"
                    self.circuit_open_until[prov] = time.time() + 60
                    self.health.record_healing(
                        prov,
                        model,
                        "circuit_opened",
                        prev_status=old,
                        new_status="open",
                        details=f"{c + 1} consecutive failures",
                    )

    def sticky_get(self, sid: str) -> tuple[str | None, str | None]:
        return self.health.sticky_get(sid, STICKY_TTL)

    def sticky_set(self, sid: str, p: str, m: str) -> None:
        self.health.sticky_set(sid, p, m)

    def circuit_ok(self, p: str) -> bool:
        st = self.circuit.get(p, "closed")
        if st == "closed":
            return True
        if st == "open":
            if time.time() > self.circuit_open_until.get(p, 0):
                self.circuit[p] = "half"
                return True
            return False
        return True  # half-open: allow probe




state = Matrix()

File: projects/herd/mesh/router/sovereign-router-ts/router_matrix.ts
import { HealthDB } from "./router_health.ts";
import { DB_PATH, STICKY_TTL, PROVIDERS } from "./router_config.ts";

// ---------------------------------------------------------------------------
// Matrix state
// ---------------------------------------------------------------------------
export class Matrix {
  fail = new Map<string, [number, number]>();
  elo = new Map<string, number>();
  circuit = new Map<string, string>();
  circuitOpenUntil = new Map<string, number>();
  fifoDepth = 0;
  health: HealthDB;

  constructor() {
    for (const p of Object.keys(PROVIDERS)) {
      this.elo.set(p, 1000);
      this.circuit.set(p, "closed");
    }
    this.health = new HealthDB(DB_PATH);
  }

  record(
    model: string,
    prov: string,
    status: number,
    lat: number,
    winner = 0,
    strategy = "",
    session = "",
  ): void {
    const latMs = lat * 1000;
    this.health.recordRequest(
      prov,
      model,
      status,
      latMs,
      strategy,
      winner,
      session,
    );
    if (status === 200) {
      const old = this.circuit.get(prov) || "closed";
      this.elo.set(prov, (this.elo.get(prov) || 1000) + 16);
      this.fail.set(prov, [0, Date.now() / 1000]);
      this.circuit.set(prov, "closed");
      if (old !== "closed") {
        this.health.recordHealing(
          prov,
          model,
          "circuit_recovered",
          old,
          "closed",
        );
      }
    } else if (status === 429) {
      this.health.recordRateLimit(prov, model, status);
      this.elo.set(prov, Math.max(100, (this.elo.get(prov) || 1000) - 8));
      const [c] = this.fail.get(prov) || [0, 0];
      this.fail.set(prov, [c + 1, Date.now() / 1000]);
    } else {
      const [c] = this.fail.get(prov) || [0, 0];
      this.fail.set(prov, [c + 1, Date.now() / 1000]);
      this.elo.set(prov, Math.max(100, (this.elo.get(prov) || 1000) - 32));
      if (c + 1 >= 3) {
        const old = this.circuit.get(prov) || "closed";
        this.circuit.set(prov, "open");
        this.circuitOpenUntil.set(prov, Date.now() / 1000 + 60);
        this.health.recordHealing(
          prov,
          model,
          "circuit_opened",
          old,
          "open",
          `${c + 1} consecutive failures`,
        );
      }
    }
  }

  stickyGet(sid: string) {
    return this.health.stickyGet(sid, STICKY_TTL);
  }
  stickySet(sid: string, p: string, m: string) {
    this.health.stickySet(sid, p, m);
  }

  circuitOk(p: string): boolean {
    const st = this.circuit.get(p) || "closed";
    if (st === "closed") return true;
    if (st === "open") {
      if (Date.now() / 1000 > (this.circuitOpenUntil.get(p) || 0)) {
        this.circuit.set(p, "half");
        return true;
      }
      return false;
    }
    return true; // half-open probe
  }
}

export const state = new Matrix();
